### ISSUE TITLE
vercel@33.3.0 broke preBuilds in monorepo

### ISSUE THREAD

**Issue by TriPSs** [zeit/now#11097]
We are using the `preBuild` option in our monorepo but after updating to vercel@33.3.0 it no longer works, when trying to do the deploy we get the following error:
```
[debug] [2024-01-26T09:52:00.561Z] Error: Error: ENOENT: no such file or directory, lstat '/home/runner/work/<org>/<project>/dist/apps/<app-name>/node_modules/styled-jsx/index.js'

Error: ENOENT: no such file or directory, lstat '/home/runner/work/<org>/<project>/dist/apps/<app-name>/node_modules/styled-jsx/index.js'
```

When comparing it with the previous version it looks like the functions don't have the node_modules anymore. How can we solve this?

**Comment by georgegebbett** [comment]
We've seen this too - we get the below error in Vercel when trying to deploy a prebuilt app

We are using Turborepo + pnpm which seems to be a fairly tricky setup for this kind of thing - our downfall is definitely coming from installing `vercel@latest` in CI as this isn't the first time our pipeline has been broken by a bad CLI release...

![image](https://github.com/vercel/vercel/assets/35996500/50e68437-533f-478b-8e8f-574966526c9e)

**Comment by wilson-webdev** [comment]
Having the same issue with a GitHub action that uses `npx vercel`

**Comment by andyvanosdale** [comment]
It looks like [commit 2b71ee6b427996dc26547a8601cf8b65c74a2ccd]( https://github.com/vercel/vercel/commit/2b71ee6b427996dc26547a8601cf8b65c74a2ccd#diff-5b7436deccee2ee3d2737e02d670d2e27c3fa229b1fea3646e7de3d5d33bf831R113) is checking the file paths in the `.vc-config.json` files to see if they are present. However, this incorrectly assumes that the `.next` directory was built out instead of using all the pre-built files. In essence, prebuilt is completely broken as a stand-alone feature, as documented in the (vercel build)[https://vercel.com/docs/cli/build] docs:

> The vercel build command can be used to build a Vercel Project locally or in your own CI environment. Build artifacts are placed into the .vercel/output directory according to the [Build Output API](https://vercel.com/docs/build-output-api/v3).
> 
> When used in conjunction with the vercel deploy --prebuilt command, this allows a Vercel Deployment to be created without sharing the Vercel Project's source code with Vercel.

**Comment by andyvanosdale** [comment]
CC @TooTallNate for the commit

**Comment by andyvanosdale** [comment]
I'll also add: it's always a good idea to version lock every dependency. This was not something we did with the Vercel CLI in our pipeline, and it took a bit of debugging to find out what went wrong. Instead of globally installing the CLI, we are now adding it as a dev dependency and calling it via `npm exec vercel -- pull/build/deploy`. I highly recommend using this approach, and I hope the Vercel team updates the documentation accordingly.

**Comment by xxxle0** [comment]
downgrade the version to 33.2.0 will work again

**Comment by EndangeredMassa** [comment]
We are looking into this. Downgrading to 33.2.0 should have you covered in the mean time.

**Comment by andyvanosdale** [comment]
Thanks, @EndangeredMassa! As y'all look into this, please note that it is not just node modules that will cause the issue. It is many built files in the `.next` directory as well.

**Comment by hussamkhatib** [comment]
> I'll also add: it's always a good idea to version lock every dependency. This was not something we did with the Vercel CLI in our pipeline, and it took a bit of debugging to find out what went wrong. Instead of globally installing the CLI, we are now adding it as a dev dependency and calling it via `npm exec vercel -- pull/build/deploy`. I highly recommend using this approach, and I hope the Vercel team updates the documentation accordingly.

@andyvanosdale  Wouldn't installing globally with specific version do the job ?

**Comment by andyvanosdale** [comment]
Not necessarily. It may get semver’d to update to minor versions, like what happened here.
________________________________

**Comment by jakeleventhal** [comment]
Does anyone have a repro of this?

**Comment by hussamkhatib** [comment]
@jakeleventhal 
~~You can add the [.github/workflows/preview.yaml](https://vercel.com/guides/how-can-i-use-github-actions-with-vercel) to any of your nextjs projects to reproduce it.~~ 
I'll create a repo with steps to reproduce the issue. It can be directly reproduced with cli, Github actions is not required.

**Comment by TooTallNate** [comment]
@hussamkhatib I can not reproduce with that. Can you show the _exact_ steps please?

**Comment by hussamkhatib** [comment]
Sure @TooTallNate 

- Clone this [repo](https://github.com/hussamkhatib/vercel-cli-gh-action-issue/tree/main), preferably in the same directory as [vercel](https://github.com/vercel/vercel)
- Open 2 terminals in the root dir of both the repo's 
   - 1st  `cd vercel-cli-gh-action-issue && git checkout no-db-gh-action`  // ignore the branch and repo name , gh actions is not relevant here.
   - 2nd `cd vercel/packages/cli` // assuming initial setup is already in place pnpm install etc
From the 2nd terminal i.e vercel cli
- Link the repo 
`pnpm dev link --repo  --cwd  ../../../vercel-cli-gh-action-issue`
- Build 
`pnpm dev build --cwd  ../../../vercel-cli-gh-action-issue/apps/web` // you will be prompted to pull projects settings etc 
- Deploy with --prebuilt
`pnpm dev deploy --prebuilt --cwd  ../../../vercel-cli-gh-action-issue/apps/web`
   
 
 --- 
Other notes 
- Vercel 33.6.1 was used at the time of testing 
- This issue might have something to do with it being a monorepo, as I tried to reproduce this in a fresh project which is not a monorepo where I couldn't reproduce it. 
- I don't if it is a must or not, I had created a project in vercel dashboard for `apps/web`  
- I am not sure if this is case with next.js app router as well.

**Comment by TooTallNate** [comment]
For anyone who has been bitten by this bug, can I ask you to try installing this version of the CLI and let me know if that fixes it for you?

```
npm i https://vercel-git-tootallnate-zero-1527-bug-we-should-look-into-9b5350.vercel.sh/tarballs/vercel.tgz
```

**Comment by sanderkranz** [comment]
@TooTallNate no, for me it doesn't. It still tries to take local files from node_modules.

<img width="799" alt="image" src="https://github.com/vercel/vercel/assets/4821962/7a19c944-3d4a-4caa-bb48-a28519e40b1e">

Just to be clear. In our case we pre-build application using `vercel build --prod` and archive `.vercel/output` folder. Later we unarchive `.vercel/output` folder on another server and run `vercel deploy --prod --prebuilt` command. I's working fine with vercel@33.2.0.

**Comment by jakeleventhal** [comment]
@TooTallNate in my case, this version (results in a MUCH faster upload)

**Comment by jakeleventhal** [comment]
```
▶ npx vercel deploy --prebuilt --archive=tgz
Vercel CLI 33.6.1-c35784f
⠇ Uploading [===-----------------] (43.7MB/284.3MB)
```
without `--archive=tgz` (seemingly no longer needed) its only about 29 MB and is able to leverage build cache

**Comment by TooTallNate** [comment]
@sanderkranz Yes, this is expected. What is the issue? And just to confirm, are you using `vc link --repo`?

**Comment by sanderkranz** [comment]
@TooTallNate No, we do not link our repo to Vercel project. So, the goal we were trying to achieve is to create the build once (calculate the checksum) and then reuse it for deployment at any time so that make sure the code remains unchanged. This may not be the best way to use Vercel, but this is how it works for now.

**Comment by p5** [comment]
:+1: 
This is our workflow, which is not working with the development build of the CLI:
1. Install Vercel CLI & install NPM deps
2. Run `vercel pull --environment production --token <token>`
3. Run `vercel build --prod --token <token>`
4. Run `vercel deploy --prebuilt --token <token>`

We don't pass the built artefacts around, unlike sanderkranz, so I don't think this is causing the issue.  
Since we are not using any Vercel/GitHub integrations with our repo, we are not linking the projects.

Something that may be worth noting, we are using Yarn workspaces in our monorepo.

To add to the benefits, we do not need to pay for additional concurrent runners in Vercel since builds are all done in our environment, plus we can efficiently cache dependencies.  Deployments happen MUCH faster when we do it this way, using Vercel as purely a hosting environment, rather than CD and hosting.

**Comment by TooTallNate** [comment]
@sanderkranz @p5 I'm not sure I understand. Would you be able to share your workflow files or a reproduction repo? I was able to speak with @jakeleventhal on a call to understand their setup, but would be happy to do the same with you.

**Comment by p5** [comment]
@TooTallNate Please see the builds in this repo for a minimal reproduction of this issue

The repo contains a create-next-app in a Yarn Workspaces monorepo, with the web project being built in CI and the final artefacts deployed as a prebuilt app to Vercel.

Repo: https://github.com/rsturla/vercel-sample
Build (latest CLI): https://github.com/rsturla/vercel-sample/actions/runs/8382923816/job/22957654618#step:7:20
Build (development CLI): https://github.com/rsturla/vercel-sample/actions/runs/8382923816/job/22957654779#step:7:20

Have signed [this commit](https://github.com/rsturla/vercel-sample/commit/7f1cd0613310a56513008afa184bd8f015148f94) with my email if you want to setup a chat.  Based in UK.

**Comment by TooTallNate** [comment]
I took a look at your example, thanks for that. In your case, you are misusing the CLI and likely have a misconfiguration on your project's settings on Vercel:

* The CLI commands are meant to be invoked from the root of the monorepo, _not_ the project directory
* You must set the Root Directory setting in your project's settings on Vercel (in your case the value would be `services/my-app`)

You can see on [my fork of your repo](https://github.com/TooTallNate/vercel-gh-11097) that it is deploying correctly with these changes to your workflow file:

```diff
diff --git a/.github/workflows/vercel-deploy.yml b/.github/workflows/vercel-deploy.yml
index c27364e..fa54e5f 100644
--- a/.github/workflows/vercel-deploy.yml
+++ b/.github/workflows/vercel-deploy.yml
@@ -28,20 +28,16 @@ jobs:
           node-version: 20
 
       - name: Install dependencies
-        working-directory: ./services/my-app
         run: |
           yarn install --immutable
           npm install -g ${{ matrix.vercel-cli-version }}
           vercel --version
 
       - name: Pull Vercel project
-        working-directory: ./services/my-app
         run: vercel pull --environment production --token ${{ secrets.VERCEL_TOKEN }}
 
       - name: Build project
-        working-directory: ./services/my-app
         run: vercel build --prod --token ${{ secrets.VERCEL_TOKEN }}
 
       - name: Deploy project
-        working-directory: ./services/my-app
         run: vercel deploy --prod --prebuilt --token ${{ secrets.VERCEL_TOKEN }}

```

**Comment by p5** [comment]
I see.  The issue was that the deployment worked with the old version (v33.2.0) and not any later releases, so I wasn't aware I was doing something wrong ([working build](https://github.com/rsturla/vercel-sample/actions/runs/8383339700/job/22958864007)).  

~~So rather than setting the working-directory to be the project directory, we should use `vercel --cwd <directory>` instead from the Yarn (or NPM) workspace root if we do not wish to manually update configuration in the Vercel UI?~~

Edit: Using `--cwd` caused the same error.
We don't like setting any configuration in the UI since it's not reproducible or versioned, hence we make good use of the vercel.json file when declaring options like the build commands or framework.  **If we also need to declare the location of the project, how can we do that declaratively?**  In v33.2.0, we just ran the commands in the project directory, but that's no longer an option.

I know this probably seems similar to [the xkcd comic](https://xkcd.com/1172), but I don't believe what myself, [hussamkhatib](https://github.com/hussamkhatib) and likely the majority of others in this thread are trying to do is anything too unthinkable.

**Comment by jakeleventhal** [comment]
@TooTallNate should we expect this to be live in the next release (>33.6.2)?

**Comment by TooTallNate** [comment]
I wanted to follow up here. So there were two reported configurations that were causing this error:
 1. When the project was linked via `vercel link --repo` (@jakeleventhal). This was a bug in Vercel CLI which has been fixed in https://github.com/vercel/vercel/pull/11309.
 2. Misusage of the `vercel build` / `vercel deploy --prebuilt` commands by running them in the wrong directory. The commands need to be run in the _root_ of the repository, as opposed to the project-specific subdirectory.

@p5 I wanted to address the use-case you described, because believe me, I feel you about wanting to keep all configuration checked into version control. It's true that you can not configure the `rootDirectory` in the `vercel.json` file because that file is placed within that project directory, so there's a chicken and egg problem. Given that, I think what you are trying to do (by specifying that value as a command line argument) _does make sense_ from a conceptual standpoint. Unfortunately it's not how the `--cwd` flag is intended to be used, in the current implementation at least.

We've been discussing your use-case internally and, given that there is a "workaround" of configuring the Root Directory setting in Project Settings, we won't be addressing the in-code config at this time. If it were added though, it should probably be a new flag (`--root-directory`) as opposed to overloading the `--cwd` flag. A pull request would be accepted for that if someone is interested.

**Comment by TooTallNate** [comment]
And thanks for the xkcd reference. Definitely relevant 😆

**Comment by sanderkranz** [comment]
> Misusage of the vercel build / vercel deploy --prebuilt commands by running them in the wrong directory. The commands need to be run in the root of the repository, as opposed to the project-specific subdirectory.

@TooTallNate, does it mean both commands have to be run in the directory containing source code? What is the use case? In my opinion, `vercel build` should generate a compiled version of application so that later we can take it and deploy anywhere via `vercel deploy --prebuild`. And it was working fine in 33.2.0, we were storing `.vercel/output` directory only. But now, how I understand, we should store node_modules directory as well. Does this case make sense or maybe you have a recommendation how to change the flow?

**Comment by jakeleventhal** [comment]
@TooTallNate this isssue is not resolved for me on `vercel@33.6.3`

GH actions yaml:
```yaml
- name: Deploy to Vercel
  shell: bash
  run: |
    preview_url=$(npx vercel --prebuilt ${{ env.VERCEL_ENV_FLAG }} --token=${{ inputs.vercel-token }} --cwd=${{ inputs.working-directory }})
    if [[ -z "${{ inputs.alias }}" ]]; then
      echo "No alias found"
    else
      npx vercel alias $preview_url ${{ inputs.alias }} --token=${{ inputs.vercel-token }} --cwd=${{ inputs.working-directory }}
    fi
  env:
    VERCEL_ORG_ID: ${{ inputs.vercel-org-id }}
    VERCEL_PROJECT_ID: ${{ inputs.project-id }}
```

Output in GH action:
```
Run preview_url=$(npx vercel --prebuilt '' --token=*** --cwd=path/to/app)

preview_url=$(npx vercel --prebuilt '' --token=*** --cwd=path/to/app)
if [[ -z "www.my.domain.com" ]]; then
  echo "No alias found"
else
  npx vercel alias $preview_url www.my.domain.com --token=*** --cwd=path/to/app
fi
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  NODE_OPTIONS: --max-old-space-size=8192
  IS_PRODUCTION: false
  ALIAS: www.my.domain.com
  PNPM_VERSION: 8.15.6
  PNPM_HOME: /home/runner/setup-pnpm/node_modules/.bin
  VERCEL_ENV: preview
  VERCEL_ENV_FLAG: ''
  VERCEL_ORG_ID: ***
  VERCEL_PROJECT_ID: ***

Vercel CLI 33.6.3
Retrieving project…
The vercel.json file should be inside of the provided root directory.
Deploying myorg/my-app
Error: Please ensure project dependencies have been installed:
File does not exist: "node_modules/.pnpm/styled-jsx@5.1.1_@babel+core@7.23.3_react@18.2.0/node_modules/styled-jsx/index.js"
Error: Process completed with exit code 1.
```

FWIW, locally when I run this, it seems to work.

My workflow before `vercel@33.3.0` was different in GH action. See diff for how I changed it for 33.6.3
<img width="1446" alt="Screenshot 2024-04-03 at 2 52 33 PM" src="https://github.com/vercel/vercel/assets/15133828/0d0128d6-1d83-4efe-a37a-46f32ad8a3c5">

**Comment by TriPSs** [comment]
I just updated to the latest version but I'm still getting: 
```
> [debug] [2024-04-05T20:51:18.742Z] Error: Error: Please ensure project dependencies have been installed:
File does not exist: "node_modules/styled-jsx/index.js"
```

Can we reopen this ticket?

**Comment by TooTallNate** [comment]
I reopened because @jakeleventhal is still having issues which we are debugging in private.

For anyone else still having this issue, you'll need to share more information about your set up, such as workflow files. Also feel free DM me on Twitter / Discord if you're more comfortable doing it that way.

**Comment by TooTallNate** [comment]
Jake's issue has been sorted out at this point. If nobody else it going to share their workflows, then I will close this issue.

**Comment by TriPSs** [comment]
We have the issue locally, what we do is the following:
1. `npx vercel pull --yes --environment=production --debug`
2. `npx vercel build --output dist/<project dist>/.vercel/output --prod --debug`
3. `npx vercel deploy --prebuilt --prod --regions=fra1 --debug`, this one is run inside the `dist/<project dist>`
3.a `npx vercel deploy <path> --prebuilt --prod --regions=fra1 --debug`, ran in the root creates the same result.

**Comment by sanderkranz** [comment]
@TooTallNate thanks for your time. Here is our workflow (simplified).
When we are releasing a new version, we build our application and attach `.vercel/output` files to GitHub release.
```
- name: 'Pull Vercel project settings'
  run: vercel pull --environment=production --yes --token=${{ secrets.VERCEL_TOKEN }} --scope=${{ env.VERCEL_SCOPE }}

- name: 'Build project artifacts'
  run: |
    vercel build --prod --token=${{ secrets.VERCEL_TOKEN }} --scope=${{ env.VERCEL_SCOPE }}
    (cd .vercel/output && tar -czf ../../build.tar.gz .)

- name: 'Release'
  run: semantic-release
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

Later on we take the attachment from GitHub release and deploy it to Vercel.
```
- name: 'Download asset from a specific release version'
  uses: robinraju/release-downloader@v1.8
  with:
    tag: v${{ github.event.inputs.version }}
    fileName: v${{ github.event.inputs.version }}.tar.gz

- name: 'Unpack build archive'
  run: mkdir -p .vercel/output && tar -xzf v${{ github.event.inputs.version }}.tar.gz -C .vercel/output

- name: 'Install Vercel CLI'
  run: npm install --global vercel@33.2

- name: 'Deploy project artifacts to Vercel'
  id: 'vercel-deploy'
  run: |
    deploymentUrl=$(vercel deploy --prod --prebuilt --skip-domain --archive=tgz --token=${{ secrets.VERCEL_TOKEN }} --scope=${{ env.VERCEL_SCOPE }})
    echo "deployment-url=$deploymentUrl" >> $GITHUB_OUTPUT

- name: 'Assign domains'
  run: vercel alias set ${{ steps.vercel-deploy.outputs.deployment-url }} ${{ env.PRE_PRODUCTION_HOST }} --token=${{ secrets.VERCEL_TOKEN }} --scope=${{ env.VERCEL_SCOPE }}
```

**Comment by hussamkhatib** [comment]
Off topic: Has anyone else noticed that entire source code is uploaded to Vercel instead of just artifacts ? Bringing it up here as I think the fix for this issue is responsible for it(vercel@33.6.3).

**Comment by tdemanincor-qm** [comment]
We started running into issues 4 days ago with our GitHub Actions using the vercel@latest. I am still getting this error in the github action agent and locally when using prebuilt and the newer versions of vercel cli. deploying without prebuilt works. We are in a yarn workspace monoreop with turbo repo. Our working directory has and was set correctly in vercel web ui project config. The issue seems to be related to the api pages and the handling of the runtime. The error we get in the agent and locally now is `Error: The following Serverless Functions contain an invalid "runtime":  - api/health (nodejs18.x)` The project runtime is set to node 18.x, the project.json nodeVersion is 18.x, the local and agent node version is 18. works without prebuilt and works with vercel@33.2.0.

**Comment by cb1kenobi** [comment]
UPDATE: I was able to solve this by unsetting the `VERCEL_ORG_ID` and `VERCEL_PROJECT_ID` env vars that I defined as GitHub env secrets, then adding a new step before deploying:

```
      - name: Link Apps
        shell: bash
        run: vercel link --repo --yes --token=${{ env.VERCEL_TOKEN }}
```

I believe I was using the `VERCEL_PROJECT_ID` from one of the sites for ALL of the sites. Doh!

I tested doing `vc deploy` vs `vc build && vc deploy prebuilt` and they took about the same amount of time, however `vc deploy` is slightly cheaper than prebuilding.

--

@TooTallNate Yo! We just moved our pipeline to a GH workflow with manual build and deploy. Now we're running into this same issue.

We have a monorepo with 2 Next.js sites. Vercel CLI 34.2.0. Next.js 14.2.3. Our GH workflow builds each site in series. For each site we vc pull, vc build, and vc deploy --prebuilt.

This error happens on deploy via our GitHub workflow:

```
Error: ENOENT: no such file or directory, lstat '/node_modules/.pnpm/styled-jsx@5.1.1_react@18.3.1/node_modules/styled-jsx/index.js'
    at Now.handleDeploymentError (/usr/local/lib/node_modules/vercel/dist/index.js:141656:16)
    at processDeployment (/usr/local/lib/node_modules/vercel/dist/index.js:141414:34)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Now.create (/usr/local/lib/node_modules/vercel/dist/index.js:141544:28)
    at async createDeploy (/usr/local/lib/node_modules/vercel/dist/index.js:141138:12)
    at async deploy_default (/usr/local/lib/node_modules/vercel/dist/index.js:148666:22)
    at async main13 (/usr/local/lib/node_modules/vercel/dist/index.js:178529:18)
```

I can reproduce locally (macOS 14.4.0, Node.js 20.11.0), though the error is slightly different:

```
Error: ENOENT: no such file or directory, lstat '/node_modules/.pnpm/next@14.2.3_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/server/next-server.js'
    at Now.handleDeploymentError (/Users/chris/Library/Application Support/fnm/node-versions/v20.11.0/installation/lib/node_modules/vercel/dist/index.js:141656:16)
    at processDeployment (/Users/chris/Library/Application Support/fnm/node-versions/v20.11.0/installation/lib/node_modules/vercel/dist/index.js:141414:34)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Now.create (/Users/chris/Library/Application Support/fnm/node-versions/v20.11.0/installation/lib/node_modules/vercel/dist/index.js:141544:28)
    at async createDeploy (/Users/chris/Library/Application Support/fnm/node-versions/v20.11.0/installation/lib/node_modules/vercel/dist/index.js:141138:12)
    at async deploy_default (/Users/chris/Library/Application Support/fnm/node-versions/v20.11.0/installation/lib/node_modules/vercel/dist/index.js:148666:22)
    at async main13 (/Users/chris/Library/Application Support/fnm/node-versions/v20.11.0/installation/lib/node_modules/vercel/dist/index.js:178529:18)
```

Note: we are _not_ using `--archive=tgz`.

You're welcome to impersonate and dig into our deployment. Find me on the Slacks. Happy to test out a debug build!

**Comment by sebs** [comment]
we stumbled over this as well: 

we had a pipeline that was split in multiple steps

1. doing a build & passing the results to the next step via artifacts
2. running tests
4. deploy

So a little different from the example, where everything in done  in one step. 

This worked on our dev environment, but the customer had a hard artifact limit, so we could not pass vercel/next installed deps there. This  lead us down this rabbit hole. 

Assumption: to deploy the build we only need vercel
Reality: the dependency in question is in next.js 

Easy Solution: run everything (build etc) from source in the deploy step and ignore all the assets created. Kind of how it is done in the docs. 

Another thing to try: run a npm install (and re-use the generated sources).

**Comment by designoor** [comment]
Having the same issue: `ENOENT: no such file or directory, lstat '/vercel/node_modules/next/node_modules/styled-jsx/index.js'`

My setup is a single monorepo with subfolders deployed to different services. I'm using yarn workspaces. The folder structure looks like this:

```
- backend (standalone websockets server deployed to heroku)
- database (database schema deployed to supabase)
- frontend (standard react app deployed to vercel)
- functions (serverless functions build with Next deployed to vercel)
- scripts (helpers not deployed anywhere)
```
root package.json

```
{
"workspaces": [
    "database",
    "functions",
    "backend",
    "frontend",
    "scripts"
  ]
}
```


For vercel, I am using CLI version `^34.2.6`. And I'm using Next `14.2.3` for the serverless build. 

### frontend

- Standard react app
- When I navigate to `frontend`:
   -  ✔ I build the app (vercel build)
   - ✔ Deploy it (vercel deploy --prebuilt)
   - ✔ Everything works just fine.

```
{
  "name": "project-website",
  "version": "0.0.0",
  "private": true,
  "proxy": "http://localhost:8000",
  "scripts": {
    "start": "NODE_ENV=production webpack-dev-server --mode production --no-live-reload --config config/webpack.prod.ts",
    "dev": "NODE_ENV=development webpack-dev-server --mode development --open --hot --config config/webpack.dev.ts",
    "build": "webpack --mode production --config config/webpack.prod.ts"
  },
  "devDependencies": {
    "@babel/core": "^7.19.6",
    "@babel/preset-env": "^7.19.4",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@types/dotenv-webpack": "^7.0.3",
    "@types/node": "^20.6.5",
    "@types/react": "^18.0.27",
    "@types/react-dom": "^18.0.6",
    "@types/react-resizable": "^3.0.6",
    "@types/webpack": "^5.28.0",
    "babel-loader": "^9.1.2",
    "buffer": "^6.0.3",
    "clean-webpack-plugin": "^4.0.0",
    "copy-webpack-plugin": "^11.0.0",
    "css-loader": "^6.7.3",
    "dotenv-webpack": "^8.0.1",
    "html-webpack-plugin": "^5.5.0",
    "react-scripts": "^5.0.1",
    "ts-node": "^10.9.1",
    "tsconfig-paths-webpack-plugin": "^4.0.0",
    "typescript": "^5.2.2",
    "webpack": "^5.74.0",
    "webpack-cli": "^5.0.1",
    "webpack-dev-server": "^4.11.1",
    "webpack-merge": "^5.8.0"
  },
  "dependencies": {
    "@metamask/browser-passworder": "^4.1.0",
    "@mui/base": "^5.0.0-beta.16",
    "dayjs": "^1.11.10",
    "dexie": "^3.2.4",
    "dexie-react-hooks": "^1.1.6",
    "ethers": "^6.7.1",
    "overlayscrollbars": "^2.4.4",
    "overlayscrollbars-react": "^0.5.3",
    "re-resizable": "^6.9.11",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-draggable": "^4.4.6",
    "react-resizable": "^3.0.5",
    "react-use-websocket": "^4.8.1",
    "siwe": "^2.1.4",
    "viem": "^1.16.6"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

### functions

NextJS project using only API routes.

- When I navigate to `functions`.
   -  ✔ I build the app (vercel build)
   - ❌ Try deploying it (vercel deploy --prebuilt, using it without --prebuilt makes no difference in my case) and I get the `no such file or directory` error.
```
// next.config.mjs

/** @type {import('next').NextConfig} */

const nextConfig = {
  transpilePackages: ['../database'],
};

export default nextConfig;
```


```
{
  "name": "project-functions",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "deploy": "next build && yarn vercel deploy --prebuilt"
  },
  "dependencies": {
    "dotenv": "^16.4.5",
    "drizzle-orm": "^0.30.10",
    "ethers": "^6.12.1",
    "iron-session": "^8.0.1",
    "next": "^14.2.4",
    "nextjs-cors": "^2.2.0",
    "oauth-1.0a": "^2.2.6",
    "postgres": "^3.4.4",
    "react": "^18",
    "react-dom": "^18",
    "siwe": "^2.3.2",
    "zod": "^3.23.6",
    "zod-validation-error": "^3.2.0"
  },
  "devDependencies": {
    "@types/node": "^20.12.12",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "typescript": "^5.4.5",
    "vercel": "^34.2.6"
  }
}
```

## I tried

- Reinstalling all the packages
- Upgrading to the latest versions
- Running the command from root and nested in functions. None worked.
- Installing `styled-jsx` dependency even though `functions` the project is not using it
- Out of curiosity tried downgrading to `33.2.0` but it failed as well.

**Comment by psychobolt** [comment]
If you guys are using `yarn` and your monorepo setup relies on a **node-linker** package that is not `node-modules`, you'll be having problems. If you need to build your other packages prior to the app, here is what you can do if you want to configure it all in Vercel's install command: `YARN_NODE_LINKER=<pnp|pnpm> yarn install && YARN_NODE_LINKER=<pnp|pnpm> yarn turbo run build-packages && yarn install` . So basically the process is, you'd first install without Vercel's `YARN_NODE_LINKER` override and build other packages that your Vercel app relies on prior to the last command is what reverts all linking. I guess for some setups, its all within one build command which can cause incompatibility with the CLI...

**Comment by reubendrummond** [comment]
I'm running into very similar errors as @0xryowa [here](https://github.com/vercel/vercel/issues/11097#issuecomment-2163250942) trying to deploy next js app. 

Using turborepo monorepo with pnpm@9. Have tried vercel@34.3.1 and 33.2.0. 

Deploying with prebuild with vercel@34.3.1 fails
```
Error: ENOENT: no such file or directory, lstat '/node_modules/.pnpm/styled-jsx@5.1.1_@babel+core@7.23.6_react@18.2.0/node_modules/styled-jsx/index.js'
```

Deploy works with vercel@33.20, but get the runtime import error
```
Cannot find module 'next/dist/server/next-server.js'
Require stack:
- /var/task/___next_launcher.cjs
Did you forget to add it to "dependencies" in `package.json`?
```

Deploying directly without prebuild doesn't work since I'm not uploading from the root and I want to avoid needing to configure the vercel project as outlined [here](https://github.com/vercel/vercel/issues/11097#issuecomment-2033540402). I don't want to deploy from root since I want each package to define their own deploy scripts. 

I've tried standalone builds and defining `outputFileTracingIncludes` in `next.config.mjs`. 

I've spent a lot of time on this and there doesn't seem to be an answer for this. I would appreciate any help. Thanks.

**Comment by osdiab** [comment]
@sanderkranz were you ever able to figure out how to your project's setup to work, with archiving the build outputs & uploading an artifact in one job, then pulling the artifact + unarchiving + deploying in another?

My company also has been doing this with no problems and great effect; I hoped to update the `vercel` CLI since at this point the old versions have a number of vulnerable dependencies, but then this same issue came up.

I also tried running `pnpm install` even though the app is already built to just have the `node_modules` directory around, but that didn't help. Frustrating.

**Comment by Pinpickle** [comment]
My workflow is with a pnpm monorepo (I've tried both on v8 and v9) and I'm facing this issue.

From within the directory that contains the next.js app I'm trying to deploy to Vercel

```
vercel build --prod
vercel deploy --prebuilt --prod
```

Which gives me the following error in Vercel CLI v35:

```
Error: ENOENT: no such file or directory, lstat '/node_modules/.pnpm/styled-jsx@5.1.1_react@18.2.0/node_modules/styled-jsx/index.js'
```


In Vercel v33.2.0, my `.vercel/output` directory has a `node_modules` directory. in v35, **it does not**. I'm guessing this is the underlying cause.

Here is my `vercel.json`:

```
{
  "buildCommand": "pnpm vercel-build-command",
  "installCommand": "echo 'Skip'",
  "framework": "nextjs"
}
```

Where `vercel-build-command` is `next build && ts-node ./scripts/deleteSourceMaps.ts`.

**Comment by lucashfreitas** [comment]
(If you have the BUILD and DEPLOY on the same jobs), I believe I have found a SOLUTION, hope it helps.

I was having the same issue on the latest version `35.2.1` and managed to fix following those steps:

1. Run everything in the root folder. Don't use `working-directory` or `--cwd` option in vercel CLI. 

2. Make sure you set the `root directory` on vercel.com dashboard: 

![image](https://github.com/user-attachments/assets/94eb9009-eaa6-4ded-8749-8ada0bdbb294)

3. This should be enough for the build to work, please see my workflow example here: 

```yaml
      - name: Vercel Pull Screts
        run: pnpm vercel pull --yes --environment=preview --token=${{env.VERCEL_TOKEN}}
      - name: Vercel Build
        id: vercel-build
        run: vercel build --token=${{env.VERCEL_TOKEN}}
      - name: Vercel Deploy
        id: vercel-deploy
        run: echo "url=$(pnpm vercel deploy --prebuilt  --token=${{env.VERCEL_TOKEN}})" >> "$GITHUB_OUTPUT"
```

**Comment by osdiab** [comment]
@lucashfreitas unfortunately that doesn’t solve the issues @Pinpickle , @sanderkranz and myself face since the point is that we run the builds in one runner (potentially in parallel with a lot of other tasks), and then deploy in another, passing the built application between the jobs by uploading/downloading the build artifacts.

What you’re saying works so long as you do it all on one job, which is incompatible with what myself and the above commenters are trying to achieve.

With this system I don’t really get why there even is a Vercel build command separate from deploy if the resulting build artifacts are useless for making deployments… anyway hope that there’s a workaround someone has found that works in this scenario as I would love to not lock my company’s Vercel CLI version for the rest of time.

**Comment by osdiab** [comment]
@Pinpickle based on reading the relevant PR (I think): https://github.com/vercel/vercel/pull/11060

it sounds like it’s intended that the node_modules folder would be gone from. .vercel/output and that so long as the vc config mapping files and node_modules directories are present, the relevant node_modules files should get uploaded. Will do some snooping around the Vercel output to understand why me running pnpm install and restoring the archive didn’t succeed in providing the necessary files.

**Comment by ilbertt** [comment]
> (If you have the BUILD and DEPLOY on the same jobs), I believe I have found a SOLUTION, hope it helps.
> 
> I was having the same issue on the latest version `35.2.1` and managed to fix following those steps:
> 
> 1. Run everything in the root folder. Don't use `working-directory` or `--cwd` option in vercel CLI.
> 2. Make sure you set the `root directory` on vercel.com dashboard:
> 
> ![image](https://private-user-images.githubusercontent.com/11416392/356864747-94eb9009-eaa6-4ded-8749-8ada0bdbb294.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MDA1NTAsIm5iZiI6MTcyNDQwMDI1MCwicGF0aCI6Ii8xMTQxNjM5Mi8zNTY4NjQ3NDctOTRlYjkwMDktZWFhNi00ZGVkLTg3NDktOGFkYTBiZGJiMjk0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDA4MDQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlMjA2NGVlMzVhN2U3YzA4NDFiODhhZWI1ZDI1NzE2YzVjMGFhOTBkNWY2MGYyOWQxNmUzNTdmZDQ3ZGVjOWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-2X75MBFJukPKE0BeIkn9HvBs-Jt3fL1YhCIGy1B0Ko)
> 
> 3. This should be enough for the build to work, please see my workflow example here:
> 
> ```yaml
>       - name: Vercel Pull Screts
>         run: pnpm vercel pull --yes --environment=preview --token=${{env.VERCEL_TOKEN}}
>       - name: Vercel Build
>         id: vercel-build
>         run: vercel build --token=${{env.VERCEL_TOKEN}}
>       - name: Vercel Deploy
>         id: vercel-deploy
>         run: echo "url=$(pnpm vercel deploy --prebuilt  --token=${{env.VERCEL_TOKEN}})" >> "$GITHUB_OUTPUT"
> ```

This worked for me, in a Turborepo with pnpm.

However, it feels bit annoying that in a project that uses all tools built by Vercel itself (Turborepo and Next.js), we still need these kind of workarounds.

**Comment by ymolists** [comment]
Hi folks.

I am also facing this in a monorepo using pnpm.

```here is what we do
 vercel --version
(node:143876) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:143876) [DEP0060] DeprecationWarning: The `util._extend` API is deprecated. Please use Object.assign() instead.
Vercel CLI 33.6.1
33.6.1

// from root of monorepo
vercel build ./apps/app-site
vercel deploy --prebuilt  ./apps/app-site
same error during deploy

// from ./apps/app-site dir
vercel build 
vercel deploy --prebuilt 
same error during deploy


```

**Comment by jeffsee55** [comment]
> passing the built application between the jobs by uploading/downloading the build artifacts.

@osdiab  Are you able to confirm that the build artifacts that the deploy action receives has the `.vercel` folder and the `node_modules`?

Also, out of curiousity what's the purpose of splitting `vercel build` from `vercel deploy --prebuilt`?

**Comment by osdiab** [comment]
@jeffsee55 the reason is that:

- our app takes several minutes to build, so we build it first while doing a ton of other stuff in parallel, including
	- separate type check/lint/formatting tasks that check our whole monorepo at once, not just the NextJS web app
	- spinning up a full environment including a fork of our database (using Neon) and a clone of our API servers (using Hasura Cloud), applying migrations and configuration changes to those
- we collect the URLs and connection secrets for each of these dependencies, and provide them to our NextJS app as environment variables when we call the `vercel deploy --prebuilt` command
	- the above potentially takes time, so we don't want to delay building the app before the other stuff gets finished.
- in addition once e2e tests on the preview environment are complete, if it's a production deploy we deploy to our production infra as well - so we have our production build also kicked off from the start so that we don't need to wait for that to build after all of our preview deployments and e2e testing concludes.

if we naively don't parallelize building the app while the rest of CI runs it would increase our CI time by 10+ minutes, which is a big deal for our velocity.

**Comment by osdiab** [comment]
To the reason it failed:

- I am using https://github.com/amondnet/vercel-action to run the deploys but I hadn't properly specified the `vercel-version` flag, turns out need that since it wasn't pulling from the already installed version declared in a nested monorepo `package.json` using `pnpm`, so the deploy command was using an old version of the CLI
- in addition to running `pnpm install` before running `vercel deploy`, I also found I needed the `.next/` build output to be stored during the build job and restored in our deploy job as well, since the vercel deploy would fail because `.next/build-manifest.json` was missing


once I fixed those things it worked!

**Comment by amiranvarov** [comment]
How the F* is this issue still open? It is GODDAMN CRITICAL! I can't deploy my ducking nextjs app for 2 ducking days. IS ANYONE FROM VERCEL EVEN BOTHERS TO DUCKING FIX IT?

**Comment by EndangeredMassa** [comment]
Hello @amiranvarov! I'm sorry about your deployment issues.

Can you share some details about your deployment flow and what issues you are seeing in a support case or in a thread in [Vercel Community](https://vercel.community/)?
https://vercel.com/help

**Comment by EndangeredMassa** [comment]
Locking this for clarity of future use. Vercel recommends using https://vercel.com/help to get help with the platform. You can also check out https://vercel.community directly.

**Comment by TriPSs** [comment]
@EndangeredMassa bit weird that you close this as it's a issue with the CLI tool and not with Vercel itself. 

Multiple ppl shared details [including myself](https://github.com/vercel/vercel/issues/11097#issuecomment-2048346947), would appreciate it if we can re-open this ticket until it is actually solved.

**Comment by ilbertt** [comment]
> @EndangeredMassa bit weird that you close this as it's a issue with the CLI tool and not with Vercel itself.
> 
> Multiple ppl shared details [including myself](https://github.com/vercel/vercel/issues/11097#issuecomment-2048346947), would appreciate it if we can re-open this ticket until it is actually solved.

Agree, please re-open this issue!

**Comment by TooTallNate** [comment]
Please open a new issue with specific details of your setup.
